core-rust/core-api-server/src/core_api/handlers/lts/state_account_resource_balance.rs

```diff
@@ -47,13 +49,74 @@ fn handle_lts_state_account_fungible_resource_balance_internal(
         extract_component_address(&extraction_context, &request.account_address)
             .map_err(|err| err.into_response_error("account_address"))?;
 
-    // TODO: super-inefficient implementation - change before mainnet
-    let component_dump = match dump_component_state(state_manager.store(), component_address) {
-        Ok(component_dump) => component_dump,
-        Err(err) => match component_address {
-            ComponentAddress::Account(_) => return Err(not_found_error("Account not found")),
-            ComponentAddress::EcdsaSecp256k1VirtualAccount(_)
-            | ComponentAddress::EddsaEd25519VirtualAccount(_) => {
+    let address = Address::Component(component_address);
+
+    let account_substate = {
+        let substate_offset = SubstateOffset::Account(AccountOffset::Account);
+        let loaded_substate = match read_mandatory_substate(
+            state_manager,
+            RENodeId::GlobalObject(address),
+            NodeModuleId::SELF,
+            &substate_offset,
+        ) {
+            Ok(substate) => substate,
+            Err(err) => {
+                match component_address {
+                    ComponentAddress::Account(_) => {
+                        return Err(not_found_error("Account not found"))
+                    }
+                    ComponentAddress::EcdsaSecp256k1VirtualAccount(_)
+                    | ComponentAddress::EddsaEd25519VirtualAccount(_) => {
+                        return Ok(models::LtsStateAccountFungibleResourceBalanceResponse {
+                            state_version: to_api_state_version(
+                                state_manager.store().max_state_version(),
+                            )?,
+                            account_address: to_api_component_address(
+                                &mapping_context,
+                                &component_address,
+                            ),
+                            fungible_resource_balance: Box::new(
+                                models::LtsFungibleResourceBalance {
+                                    fungible_resource_address: to_api_resource_address(
+                                        &mapping_context,
+                                        &fungible_resource_address,
+                                    ),
+                                    amount: to_api_decimal(&Decimal::zero()),
+                                },
+                            ),
+                        })
+                    }
+                    _ => {
+                        // This should not happen
```

Comment:
I believe this can happen - we've only validated that the address is a component address, not an account address on line 48. So we should either handle this at validation time or at this time - but either way, we should return a client error for wrong address, not a server error.